### ISSUE TITLE
rendering i and j with setFontVector < 17 makes characters unreadable

### ISSUE THREAD

**Issue by jose1711** [espruino/Espruino#1151]
i understand that with decreasing fontsize also the rendering quality decreases but i believe something still can be done to the characters containing a dot. here's a screenshot of 
```
g.setFontVector(16);
g.drawString("abcdefghijklm.:", 0, 5);
```
![pic](https://cloud.githubusercontent.com/assets/1406222/25360608/27a73bf8-294a-11e7-81c4-b3dd30ed58b7.png)
please note how "ij" part renders.

**Comment by gfwilliams** [comment]
Thanks. However I would say 'unreadable' is a strong word for what's shown in that picture :)

Do you have any particular thoughts on what could be done? The vector font is automatically created from a source font, so I guess there are two options:

* Change the rendering somehow - but I'm not sure how
* Actually create a new font from scratch that will scale better (perhaps moving the dot further from the rest of the letter)

I'm not sure if you're aware, but for smaller fonts there are several different bitmap fonts you can use that will render much better: https://www.espruino.com/Fonts

**Comment by jose1711** [comment]
Hello there,

the description may have sounded a bit harsh but OTOH if bitmap font does not suit your needs and you really want to be able to distinguish between i, I, | and !, well.. :-)

For the sake of completeness I rendered the example for font sizes going from 7 (which does not really make sense at all) all the way up to 19.

07
![07](https://cloud.githubusercontent.com/assets/1406222/25377240/5015a8b0-29a7-11e7-8684-f6e251b677bd.png)
08
![08](https://cloud.githubusercontent.com/assets/1406222/25377238/5007c31c-29a7-11e7-9691-c9b7461b5645.png)
09
![09](https://cloud.githubusercontent.com/assets/1406222/25377235/50045380-29a7-11e7-9f2c-210f9507d5de.png)
10
![10](https://cloud.githubusercontent.com/assets/1406222/25377237/5004e84a-29a7-11e7-8724-236f2a80f2b4.png)
11
![11](https://cloud.githubusercontent.com/assets/1406222/25377239/500acf9e-29a7-11e7-855e-75f4bb66fb7c.png)
12
![12](https://cloud.githubusercontent.com/assets/1406222/25377236/50048288-29a7-11e7-9ed5-709a332c939c.png)
13
![13](https://cloud.githubusercontent.com/assets/1406222/25377241/50199cea-29a7-11e7-906c-4a921dcb2cb8.png)
14
![14](https://cloud.githubusercontent.com/assets/1406222/25377242/5019d46c-29a7-11e7-96ce-ea4de6cdf6c0.png)
15
![15](https://cloud.githubusercontent.com/assets/1406222/25377243/501d277a-29a7-11e7-9184-58528cfa3701.png)
16
![16](https://cloud.githubusercontent.com/assets/1406222/25377245/50258a1e-29a7-11e7-99a0-a2aedc050623.png)
17
![17](https://cloud.githubusercontent.com/assets/1406222/25377244/5021afde-29a7-11e7-8ada-07e230268565.png)
18
![18](https://cloud.githubusercontent.com/assets/1406222/25377246/502d8746-29a7-11e7-8ab2-32e6bdc05516.png)
19
![19](https://cloud.githubusercontent.com/assets/1406222/25377247/50380f68-29a7-11e7-8276-c7b15a77c9cb.png)

It is interesting to see how 15 looks acceptable (although it does fail when rendering an exclamation mark in another test), but 16 does not.

To be honest, I do not know what the best solution would be. I wonder though how the font is rendered - I could see the polygons defined for each letter but could not find the process which translates them to bitmaps.

**Comment by opichals** [comment]
@gfwilliams Aside from creating a bitmap fonts of specific size I can only think of compiling FreeType library in which could be fine for board with plenty of RAM.

**Comment by gfwilliams** [comment]
Since you can mix and match fonts, I'm still really struggling to see in what circumstances you absolutely need to use a small vector font but can't just use a bitmap, with a vector font for the larger characters? What's your use case?

I guess freetype would be better, but I don't think it's realistic? I just checked the size of the 'Free Sans' TTF font on Linux. It's 1.5MB! By comparison Espruino's vector font is 5kB.

Obviously the ideal would be to use grayscale and antialiasing, but the majority of devices will have a pure black and white display (and you can get grayscale by rendering at 2x size and sampling down if you need to). I just tried with Gimp and plenty of vector fonts on the PC render really badly in 1 bit at 8px high.

The character rendering code is here: https://github.com/espruino/Espruino/blob/master/libs/graphics/graphics.c#L373

It's pretty basic, but there's a huge pressure on keeping the flash and RAM needed to a minimum so quite a lot of work has gone in to both that and the font to make that happen. If you find a better way of rendering without massively increasing the font size then I'd be really interested though.

Perhaps one option would be to just render twice as large and subsample. The following code does do a better job (for a Pico with the Nokia LCD display on it):

```
A5.write(0); // GND
A7.write(1); // VCC

// Setup SPI
var spi = new SPI();
spi.setup({ sck:B1, mosi:B10 });
// Initialise the LCD
var g = require("PCD8544").connect(spi,B13,B14,B15, function() {
  // When it's initialised, clear it and write some text
  g.clear();
  g.setFontVector(10);
  g.drawString2("abcdefghijklmnopq",0,0);
  g.flip();
});

g.fontSize = 6;
g._setFontVector = g.setFontVector;
g.setFontVector = function(s) {
  this.fontSize = s;
  this._setFontVector(s);
};
g.drawString2 = function(txt,x,y) {
  var w = this.stringWidth(txt);
  var h = this.fontSize;
  var b = Graphics.createArrayBuffer(w*2,h*2,1);
  b.setFontVector(this.fontSize*2);
  b.drawString(txt);
  var x,y;
  for (y=0;y<h;y++)
    for (x=0;x<w;x++)
      if ((b.getPixel(x*2,y*2)+
           b.getPixel(x*2+1,y*2)+
           b.getPixel(x*2,y*2+1)+
           b.getPixel(x*2+1,y*2+1))>2)
        this.setPixel(x,y);
};
```

It's still far from ideal though - it's a hard problem to solve.

**Comment by gfwilliams** [comment]
Closing this as a wontfix - I don't see any simple way around it, and playing around on desktop Linux and Gimp they appear to swap to a different line-based font under 17px high.

Realistically the best solution is just to use bitmap fonts, which are already supported and easy to use.

**Comment by gfwilliams** [comment]
Minor update: I [fixed some rounding errors in fillPoly](ed6fad079cbcba40bb3a733f9250c7369b53e594) that improve the font rendering. However these still don't fix vector fonts <17 pixels.
